### PR TITLE
[Frontend] Use request id from header

### PR DESCRIPTION
**Pull request by joerunde**
This is a small proposal to populate the request ids in the engine from the `X-Request-Id` header when running the openai server. This will benefit traceability: when users pass in the header they can correlate vLLM logs with the request that generated them, something that our users are asking for. If the header is not passed, behavior remains the same.

Something to consider is the implications of users potentially passing in the same request id multiple times- nothing seems to crash when that happens, but in that case it would be hard to distinguish logs from different requests.

### CODE REVIEW

**Review by DarkLight1337 — APPROVED** [review]
Seems reasonable to me, thanks for enabling more control over this!

**Comment by cjackal** [comment]
I don't know if it's commonsensical but I believe webservers must not be affected by malicious client.

If a (malicious or purely curious) client send 1mb of string as X-Request-Header with ~20k estimated output token size streaming chat completion, then handling this request consumes additional 20gb network bandwidth for repeatedly sending the request id to the SSE response, which sounds pretty huge workload for a poor API gateway router. Or a company using vLLM as its main LLM serving engine may keep all LLM response for a long-term (to oversee their poor workers, I'm not telling my own experience 🤣 ), then its poor DB may suffer from the arbitrarily long VARCHAR column `request_id`.

Of course there is effectively no such black client out there, but when it comes who knows...

**Comment by joerunde** [comment]
> I don't know if it's commonsensical but I believe webservers must not be affected by malicious client.
> 
> If a (malicious or purely curious) client send 1mb of string as X-Request-Header with ~20k estimated output token size streaming chat completion, then handling this request consumes additional 20gb network bandwidth for repeatedly sending the request id to the SSE response, which sounds pretty huge workload for a poor API gateway router. Or a company using vLLM as its main LLM serving engine may keep all LLM response for a long-term (to oversee their poor workers, I'm not telling my own experience 🤣 ), then its poor DB may suffer from the arbitrarily long VARCHAR column `request_id`.
> 
> Of course there is effectively no such black client out there, but when it comes who knows...

Ah, that's a really good point! Though I think that problem exists today since we set `X-Request-Id` on the response already- this change only propagates that ID onto other logs in the engine. 

> Or a company using vLLM as its main LLM serving engine...

Speaking from some experience, in this case users don't get to connect to vLLM directly, and another gateway is responsible for validating requests, rate limiting, setting headers etc.

**Comment by WangErXiao** [comment]
I think the request_id must be guaranteed to be unique since it is used as a key in many places.

File: vllm/entrypoints/openai/serving_engine.py [hunk 1]
```diff
@@ -565,6 +566,14 @@ async def _get_trace_headers(
 
         return None
 
+    @staticmethod
+    def _base_request_id(raw_request: Request,
+                         default: Optional[str] = None) -> Optional[str]:
+        """Pulls the request id to use from a header, if provided"""
+        default = default or random_uuid()
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_engine.py:573]: By default, the request ID is still created from UUID.

**Pull request merged**